text/0034-sdk-lifecycle-hooks.md

```diff
@@ -0,0 +1,125 @@
+```
+
+```
+| Session | ---> | Envelope | ---> | Transport | ---> | Sentry |
+```
+
+The SDKs provide a few ways to extend this pipeline:
+
+1. Event Processors (what Integrations use)
+2. `beforeSend` callback
+3. `beforeBreadcrumb` callback
+
+But these are all top level options in someway, and are part of the unified API as a result. This means that in certain scenarios, they are not granular enough as extension points.
+
+The following are some examples of how sdk hooks can unblock new features and integrations, but not a definitive list:
+
+- Integrations want to add information to spans when they start or finish. This is what [RFC #75 is running into](https://github.com/getsentry/rfcs/pull/75), where they want to add thread information to each span.
+- Integrations want to add information to envelopes before they are sent to Sentry.
+- Integrations want to run code on transaction/span finish (to add additional spans to the transaction, for example).
+- Integrations want to mutate an error on `captureException`
+- Integrations want to override propagation behaviour (extracing/injecting outgoing headers)
+
+# Proposal
+
+SDK hooks live on the client, and are **stateless**. They are called in the order they are registered. SDKs can opt-in to whatever hooks they use, and there can be hooks unique to an SDK.
+
+Hooks are meant to be mostly internal APIs for integration authors, but we can also expose them to SDK users if there is a use case for it.
+
+As hook callbacks are not processed by the client, they can be async functions.
+
+```ts
+// Example implementation in JavaScript
+
+type HookCallback = (...args: unknown[]): void;
+
+class Client {
+  hooks: {
+    [hookName: string]: HookCallback[];
+  };
+
+  on(hookName: string, callback: HookCallback): void {
+    this.hooks[hookName].push(callback);
+  }
+
+  emit(hookName: string, ...args: unknown[]): void {
+    this.hooks[hookName].forEach(callback => callback(...args));
+  }
+}
+```
+
+SDKs are expected to have a common set of hooks, but can also have SDK specific hooks.
+
+## Hooks
+
+These following are a set of example hooks that would unblock some use cases listed above. The names/schema of the hooks are not final, and are meant to be used as a starting point for discussion.
+
+To document and approve new hooks, we will create a new page in the develop docs that lists all the hooks, and what they are used for.
+
+`startTransaction`:
+
+```ts
+on('startTransaction', callback: (transaction: Transaction) => void) => void;
+```
+
+`finishTransaction`:
+
+```ts
+on('finishTransaction', callback: (transaction: Transaction) => void) => void;
+```
+
+`startSpan`:
+
+```ts
+on('startSpan', callback: (span: Span) => void) => void;
+```
+
+`finishSpan`:
+
+```ts
+on('finishSpan', callback: (span: Span) => void) => void;
+```
+
+`beforeEnvelope`:
+
+```ts
+on('beforeEnvelope', callback: (envelope: Envelope) => void) => void;
+```
```

Comment:
Why do we need a hook for `beforeEnvelope`? Can we make it internal at least? AFAIK this is not exposed for clients and the whole envelope impl. is marked as internal in Mobile SDKs at least.

Reply:
Yes we can make some hooks internal. I will clarify this in the RFC.